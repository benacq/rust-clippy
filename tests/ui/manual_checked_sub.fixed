#![allow(clippy::collapsible_if)]
#![warn(clippy::manual_checked_sub)]

fn positive_tests() {
    let a: u32 = 10;
    let b: u32 = 5;
    let c: u32 = 3;

    if let Some(difference_1) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let difference = "some test";
        let d = difference_1;
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let d = difference;
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let d = difference;
    } else {
        let d = a + b;
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let d = difference;
    } else if a < b {
        let d = a + b;
    }

    if let Some(decremented) = a.checked_sub(1) {
        //~^ manual_checked_sub
        let _ = decremented;
    }

    if let Some(decremented) = a.checked_sub(1) {
        //~^ manual_checked_sub
        let _ = decremented;
    }

    let mut difference = "variable initialized outside the if scope";
    if let Some(difference_1) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let some_reference = difference;
        let d = difference_1;
    }

    if let Some(decremented_1) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let decremented = "variable initialized inside the if scope";
        let d = decremented_1;
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        some_function(difference);
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        macro_rules! subtract {
            () => {
                difference
            };
        }
        let _ = subtract!();
    }

    struct Example {
        value: u32,
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        let _ = Example { value: difference };
    }

    if let Some(difference) = a.checked_sub(b) {
        //~^ manual_checked_sub
        if c > 0 {
            let _ = difference;
        }
    }

    if a >= b {
        if let Some(decremented) = c.checked_sub(1) {
            //~^ manual_checked_sub
            let _ = decremented;
        }
    }
}

fn some_function(a: u32) {}

fn negative_tests() {
    let a: u32 = 10;
    let b: u32 = 5;

    let x: i32 = 10;
    let y: i32 = 5;

    let mut a_mut = 10;

    if a_mut >= b {
        a_mut *= 2;
        let c = a_mut - b;
    }

    if a_mut > 0 {
        a_mut *= 2;
        let c = a_mut - 1;
    }

    if x >= y {
        let _ = x - y;
    }

    if a == b {
        let _ = a - b;
    }

    if a > b {
        let _ = a - b;
    }

    if 10 >= 5 {
        let _ = 10 - 5;
    }
}

fn main() {
    positive_tests();
    negative_tests();
}
