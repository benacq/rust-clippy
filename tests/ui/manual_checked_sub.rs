#![allow(clippy::collapsible_if)]
#![warn(clippy::manual_checked_sub)]

fn positive_tests() {
    let a: u32 = 10;
    let b: u32 = 5;
    let c = 3;

    // Basic subtraction inside an if
    if a >= b {
        //~^ manual_checked_sub
        let c = a - b;
    }

    // Basic subtraction inside an if with an else block
    if a >= b {
        //~^ manual_checked_sub
        let c = a - b;
    } else {
        let c = a + b;
    }

    // Basic subtraction inside an if with an else-if block
    if a >= b {
        //~^ manual_checked_sub
        let c = a - b;
    } else if a < b {
        let c = a + b;
    }

    // Decrementing inside an if condition
    if a > 0 {
        //~^ manual_checked_sub
        let _ = a - 1;
    }

    // Variable subtraction used in a function call
    if a >= b {
        //~^ manual_checked_sub
        some_function(a - b);
    }

    if a >= b {
        //~^ manual_checked_sub
        macro_rules! subtract {
            () => {
                a - b
            };
        }
        let _ = subtract!();
    }

    struct Example {
        value: u32,
    }

    if a >= b {
        //~^ manual_checked_sub
        let _ = Example { value: a - b };
    }

    // Subtraction inside a nested if, should lint
    if a >= b {
        //~^ manual_checked_sub
        if c > 0 {
            let _ = a - b;
        }
    }
}

fn some_function(a: u32) {}

fn negative_tests() {
    let a: u32 = 10;
    let b: u32 = 5;

    // Uses `.checked_sub()`, should not trigger
    let _ = a.checked_sub(b);

    // Works with signed integers (i32), should not trigger
    let x: i32 = 10;
    let y: i32 = 5;

    if x >= y {
        let _ = x - y;
    }

    // If condition does not match subtraction variables
    if a == b {
        let _ = a - b;
    }

    // a - b inside an if a > b (shouldn't lint)
    if a > b {
        let _ = a - b;
    }

    // Unsuffixed literals (shouldn't lint)
    if 10 >= 5 {
        let _ = 10 - 5;
    }

    // // Suffixed LHS, unsuffixed RHS (shouldn't lint)
    // if 10u32 >= 5 {
    //     let _ = 10u32 - 5;
    // }

    // //  Unsuffixed LHS, suffixed RHS (shouldn't lint)
    // if 10 >= 5u32 {
    //     let _ = 10 - 5u32;
    // }
}

fn main() {
    positive_tests();
    negative_tests();
}
