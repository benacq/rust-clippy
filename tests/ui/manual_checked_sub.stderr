error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:9:5
   |
LL | /     if a >= b {
LL | |
LL | |         let difference = "some test";
LL | |         let d = a - b;
LL | |     }
   | |_____^
   |
   = note: `-D clippy::manual-checked-sub` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::manual_checked_sub)]`
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference_1) = a.checked_sub(b) {
LL +
LL +         let difference = "some test";
LL +         let d = difference_1;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:15:5
   |
LL | /     if b <= a {
LL | |
LL | |         let d = a - b;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         let d = difference;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:20:5
   |
LL | /     if a >= b {
LL | |
LL | |         let d = a - b;
LL | |     } else {
LL | |         let d = a + b;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         let d = difference;
LL +     } else {
LL +         let d = a + b;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:27:5
   |
LL | /     if a >= b {
LL | |
LL | |         let d = a - b;
LL | |     } else if a < b {
LL | |         let d = a + b;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         let d = difference;
LL +     } else if a < b {
LL +         let d = a + b;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:34:5
   |
LL | /     if a > 0 {
LL | |
LL | |         let _ = a - 1;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(decremented) = a.checked_sub(1) {
LL +
LL +         let _ = decremented;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:39:5
   |
LL | /     if 0 < a {
LL | |
LL | |         let _ = a - 1;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(decremented) = a.checked_sub(1) {
LL +
LL +         let _ = decremented;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:45:5
   |
LL | /     if a >= b {
LL | |
LL | |         let some_reference = difference;
LL | |         let d = a - b;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference_1) = a.checked_sub(b) {
LL +
LL +         let some_reference = difference;
LL +         let d = difference_1;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:51:5
   |
LL | /     if a > 0 {
LL | |
LL | |         let decremented = "variable initialized inside the if scope";
LL | |         let d = a - b;
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(decremented_1) = a.checked_sub(b) {
LL +
LL +         let decremented = "variable initialized inside the if scope";
LL +         let d = decremented_1;
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:57:5
   |
LL | /     if a >= b {
LL | |
LL | |         some_function(a - b);
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         some_function(difference);
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:62:5
   |
LL | /     if a >= b {
LL | |
LL | |         macro_rules! subtract {
LL | |             () => {
...  |
LL | |         let _ = subtract!();
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         macro_rules! subtract {
LL +             () => {
LL +                 difference
LL +             };
LL +         }
LL +         let _ = subtract!();
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:76:5
   |
LL | /     if a >= b {
LL | |
LL | |         let _ = Example { value: a - b };
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         let _ = Example { value: difference };
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:81:5
   |
LL | /     if a >= b {
LL | |
LL | |         if c > 0 {
LL | |             let _ = a - b;
LL | |         }
LL | |     }
   | |_____^
   |
help: consider using `.checked_sub()`
   |
LL ~     if let Some(difference) = a.checked_sub(b) {
LL +
LL +         if c > 0 {
LL +             let _ = difference;
LL +         }
LL +     }
   |

error: manual re-implementation of checked subtraction
  --> tests/ui/manual_checked_sub.rs:89:9
   |
LL | /         if c > 0 {
LL | |
LL | |             let _ = c - 1;
LL | |         }
   | |_________^
   |
help: consider using `.checked_sub()`
   |
LL ~         if let Some(decremented) = c.checked_sub(1) {
LL +
LL +             let _ = decremented;
LL +         }
   |

error: aborting due to 13 previous errors

